model/adjuster/span_merger.go

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2018 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package adjuster
+
+import (
+	"github.com/jaegertracing/jaeger/model"
+)
+
+// MergeSpans returns an Adjuster that merges Jaeger spans with the same spanID.
+// Duplicate spans that have conflicting span.kind annotations are not merged.
+func MergeSpans() Adjuster {
+	return Func(func(input *model.Trace) (*model.Trace, error) {
+
+		IDToSpans := groupByIDs(input.Spans)
+
+		if len(IDToSpans) == len(input.Spans) {
+			return input, nil
+		}
+
+		trace := &model.Trace{}
+		trace.Warnings = input.Warnings
+		for _, spans := range IDToSpans {
+			if isMergeable(spans) {
+				trace.Spans = append(trace.Spans, mergeSpans(spans))
+			} else {
+				trace.Spans = append(trace.Spans, spans...)
+			}
+		}
+		return trace, nil
+	})
+}
+
+func groupByIDs(spans []*model.Span) map[model.SpanID][]*model.Span {
+	IDToSpans := make(map[model.SpanID][]*model.Span)
+	for _, span := range spans {
+		if spans, ok := IDToSpans[span.SpanID]; ok {
+			IDToSpans[span.SpanID] = append(spans, span)
+		} else {
+			IDToSpans[span.SpanID] = []*model.Span{span}
+		}
+	}
+	return IDToSpans
+}
+
+func mergeSpans(spans []*model.Span) *model.Span {
+	finalSpan := spans[len(spans)-1]
+	//we track the index of the final span to cover the case
+	//that we didn't receive the final span yet and don't want
+	//to duplicate the members of the last span due the fallback
+	//of using the last span as carrier when no final span is stored yet
+	completeSpanIndex := len(spans) - 1
+	for i := range spans {
+		if !spans[i].GetIncomplete() {
+			completeSpanIndex = i
+			finalSpan = spans[i]
+		}
+	}
+	warnings := []string{}
+	tags := []model.KeyValue{}
+	logs := []model.Log{}
+	references := []model.SpanRef{}
+	for i, span := range spans {
+		//merge refs, tags, logs and warnings of all spans
+		//take simple values from lastSpan
+		if span.GetIncomplete() && completeSpanIndex != i {
+			references = append(references, span.GetReferences()...)
+			tags = append(tags, span.GetTags()...)
+			logs = append(logs, span.GetLogs()...)
+			warnings = append(warnings, span.GetWarnings()...)
```

Comment:
so this would just duplicate everything as many times as the incomplete spans were flushed. We need to dedupe all of these data entries. Or, alternatively, just ignore everything but the final span.

Reply:
I am reluctant to rely on the client library doing the exact right thing for this to work. Yes, if the client lib clears all collections (refs, logs, & tags) after each flush, then the simple merge is fine, but what if it doesn't?

Reply:
Yes, agreed, but what if we pick the final span and the client would implement it the other way around, thus flushing the span and only keeping spanId, traceId and startTime? We would never see any tags. that are flushed intermediately. 

So, I think there should be a spec for client implementors on how to correctly flush intermediate spans. IMHO the client needs to handle intermediate spans correctly, doesn't matter, which path we choose or do you think there is a one-size-fits-all solution for the merger?